File: AoC2023/src/main/kotlin/Day01.py
import regex as re


def get_value(element: str) -> int:
    try:
        return int(element)

    except ValueError:
        if element == "one":
            return 1
        elif element == "two":
            return 2
        elif element == "three":
            return 3
        elif element == "four":
            return 4
        elif element == "five":
            return 5
        elif element == "six":
            return 6
        elif element == "seven":
            return 7
        elif element == "eight":
            return 8
        elif element == "nine":
            return 9
        else:
            raise RuntimeError(f"CANT PARSE NUMBER: {element}")


def solve_part2() -> int:
    lines = open("D:\Workspace\AdventOfCode\AoC2023\src\main\\resources\day01.txt").readlines()
    total = 0

    for line in lines:
        match = re.findall(r'one|two|three|four|five|six|seven|eight|nine|\d', line, overlapped=True)
        first_dig = get_value(match[0])
        last_dig = get_value(match[len(match) - 1])

        total += (first_dig * 10) + last_dig

    return total


def main():
    print(f"sum is {solve_part2()}")


if __name__ == "__main__":
    main()

File: AoC2023/src/main/kotlin/Day01.kt
import java.io.File

/*
 * Copyright (c) 2023. by Dominic Gernert
 */

fun main() {
    val day01 = Day01(File("src/main/resources/day01.txt").bufferedReader().readLines())

    println("Part 1 - sum is ${day01.solvePart1()}")
    println("Part 2 - sum is ${day01.solvePart2()}")
}

class Day01(input: List<String>) {
    private val calibrationDoc = input

    fun solvePart1(): Int = calibrationDoc.sumOf { line ->
        var firstDig = -1
        var lastDig = -1

        for (i in 0..line.lastIndex) {
            if (firstDig == -1 && line[i].isDigit())
                firstDig = line[i].digitToInt()

            if (lastDig == -1 && line[line.lastIndex - i].isDigit())
                lastDig = line[line.lastIndex - i].digitToInt()

            if (firstDig != -1 && lastDig != -1)
                break
        }

        (firstDig * 10) + lastDig
    }

    fun solvePart2(): Int = calibrationDoc.sumOf { line ->
        val pattern = Regex("""one|two|three|four|five|six|seven|eight|nine|\d""")
        val r = pattern.findAllWithOverlap(line).toList()

        val firstDig = getValue(r.first().value)
        val lastDig = getValue(r.last().value)

        (firstDig * 10) + lastDig
    }

    fun getValue(input: String): Int {
        try {
            return input.toInt()
        } catch (e: NumberFormatException) {
            return when (input) {
                "one" -> 1
                "two" -> 2
                "three" -> 3
                "four" -> 4
                "five" -> 5
                "six" -> 6
                "seven" -> 7
                "eight" -> 8
                "nine" -> 9
                else -> throw RuntimeException("CANT PARSE NUMBER: $input")
            }
        }
    }
}